auth/src/main/java/io/grpc/auth/OAuth2ChannelInterceptor.java

```diff
@@ -39,40 +39,43 @@
 import io.grpc.ClientInterceptors.ForwardingCall;
 import io.grpc.Metadata;
 import io.grpc.MethodDescriptor;
+import io.grpc.Status;
 
+import java.io.IOException;
 import java.util.concurrent.Executor;
 
-import javax.inject.Provider;
-
-/** Client interceptor that authenticates all calls with OAuth2. */
+/**
+ * Client interceptor that authenticates all calls with an OAuth2 credential.
+ * Uses the new and simplified Google auth library:
+ * https://github.com/google/google-auth-library-java
+ */
 public class OAuth2ChannelInterceptor implements ClientInterceptor {
   private static final Metadata.Key<String> AUTHORIZATION =
       Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER);
 
-  private final OAuth2AccessTokenProvider accessTokenProvider;
-  private final Provider<String> authorizationHeaderProvider
-      = new Provider<String>() {
-        @Override
-        public String get() {
-          return "Bearer " + accessTokenProvider.get();
-        }
-      };
+  private final Credentials credentials;
 
-  public OAuth2ChannelInterceptor(Credential credential, Executor executor) {
-    this.accessTokenProvider = new OAuth2AccessTokenProvider(credential, executor);
+  public OAuth2ChannelInterceptor(Credentials credentials, Executor executor) {
+    // TODO(louiscryan): Need executor for supporting async access-token refresh once
+    // library supports it.
+    this.credentials = credentials;
   }
 
   @Override
   public <ReqT, RespT> Call<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
-      Channel next) {
+                                                       Channel next) {
     // TODO(ejona86): If the call fails for Auth reasons, this does not properly propagate info that
     // would be in WWW-Authenticate, because it does not yet have access to the header.
     return new ForwardingCall<ReqT, RespT>(next.newCall(method)) {
       @Override
       public void start(Listener<RespT> responseListener, Metadata.Headers headers) {
-        headers.put(AUTHORIZATION, authorizationHeaderProvider.get());
-        super.start(responseListener, headers);
+        try {
+          headers.put(AUTHORIZATION, credentials.getRequestMetadata().get("Authorization").get(0));
+          super.start(responseListener, headers);
```

Comment:
I think you are going to need to instead re-throw the IOException as a RuntimeException. Otherwise if it fails and someone tries to send a message they will get a IllegalStateException.

If you throw, the stubs will manage cleanup.